### PR TITLE
Refs #21312 - Use request hostname for webpack dynamic imports

### PR DESCRIPTION
**Pull request by tbrisker**

### CODE REVIEW

**Review by sharvit — DISMISSED** [review]

**Review by ohadlevy — DISMISSED** [review]
:+1: I like it that its simplify the code, and effect only the non production flows. I'm pretty certain that if it breaks development again we can improve :) or move to a webpack dev server as a proxy.

I've also confirmed this works in production.

**Review by amirfefer — COMMENTED** [review]
works locally but it doesn't work on `centos7-devel`, getting this:
```
bootstrap be0a6ff11f212b7937da:766 Refused to load the script 'http://0.0.0.0:3808/webpack/react-
intl/locale/en.js' because it violates the following Content Security Policy directive: "script-src 'unsafe-
eval' 'unsafe-inline' 'self' 192.168.121.17:3808".
```

**Comment by sharvit** [comment]
@amirfefer how exactly do you run the server and how do you access it? (there are too many options 😞 )

**Comment by masayag** [comment]
Looks good to me: 
I'm running foreman on dev-setup with `BIND=192.168.111.1 bundle exec foreman start`

and the address is set to the bound address:
```
14:25:12 webpack.1 | Project is running at http://192.168.111.1:3808/
14:25:12 webpack.1 | webpack output is served from /webpack/
```

**Comment by tbrisker** [comment]
@ohadlevy can you please merge? even if this doesn't cover 100% of the cases, i think it's better than having develop broken in 100% of the cases

**Comment by ohadlevy** [comment]
@tbrisker can you fix the linting issues please?

**Comment by tbrisker** [comment]
fixed

**Review by ares — COMMENTED** [review]
Does that mean that --public and --https options for dev server will be ignored again?

if that's the case, you'll break my env, this was why I asked @tstrachota to parse --public webpack dev server parameter, the setup is nginx with ssl proxying to localhost puma and webpack-dev-server

**Comment by ohadlevy** [comment]
@ares based on the fact it broke most common usage cases, i am going to merge this to unblock folks, I think we need to do the following:

1. write down all the usage cases (perhaps add basic tests?)
2. revisit which one can we eliminate, i think the pain point here is that there too many ways to do very similar things.
3. consider changing our setup, so webpack is a proxy in front of rails, (and in forklift usage case, its between apache and rails for dev env), this will hopefully reduce the amount of url parsing/guessing we need to do, and reduce complexity (e.g. having webpack on different ports etc)

**Review by ohadlevy — APPROVED** [review]

**Comment by ares** [comment]
Makes sense to unblock. Who owns items listed above? It feels we are not using webpack dev server properly, if we can't use its options.

**Comment by tbrisker** [comment]
> Does that mean that --public and --https options for dev server will be ignored again?

no, webpack dev server seems to handle these options on the command line without needing extra config, at least from the testing we have done on this. 
The problem is we have at least 4 different ways people run their dev setups (5 if i understand correctly regarding yours), and who knows how many others are out there. People also access their server in different ways - by ip, by hostname or by `localhost`. This is simply unsupportable - we need to have one way of running and accessing it, that properly handles things like pry and ssl, and doesn't care if you run it locally, in forklift or elsewhere. 

TBH i'm not sure the dynamic import for the i18n files is worth all the effort and issues it caused. On my dev setup (locally, with FF), it actually seems to cause a performance degradation of about 1 second while the webpacked code is parsed and fetches the extra file before it even starts rendering the page.

**Comment by tbrisker** [comment]
even with the files cached, it seems to take over a second before the request for the en.js file is sent and about 2s before page load:
image

**Comment by ohadlevy** [comment]
can you try in production vs development?

**Comment by tbrisker** [comment]
in production, with compiled assets, everything cached and webpack dev server disabled, it is slightly better but still about 0.5s delay:
image

**Comment by ohadlevy** [comment]
maybe this is a side effect that our vendor is that large, i wonder if it
would make it a lot faster if we break just the i18n initialization code to
another bundle (or reduce the size of vendor)

**Comment by ares** [comment]
> no, webpack dev server seems to handle these options on the command line without needing extra config, at least from the testing we have done on this.

actually it does not work again and it seems the previous workaround of setting ASSET_PATH no longer works either (which seemed as a hack and reimplementation of `--public` anyway),

I have now no idea how to configure a dev setup which uses FQDN with web server in front (nginx/apache - as close to production env as I can imagine), it always seems to try fetching the manifest from http://localhost:3808/webpack/manifest.json

EDIT: I'll need to check in more details, quick revert of this didn't solve the ASSET_PATH change

**Comment by ares** [comment]
There are couple of problems that I managed to workaround, now I'm getting error when loading the intl
```
bootstrap 4c3de8858d40fb794424:108 GET https://foreman.example.tst/webpack/react-intl/locale/en-7d31511c86c2212a45fd.js net::ERR_ABORTED 404 (Not Found)
```

not that the URL is missing the port 3808, the manifest is downloaded find using the right port, ideas?

File: webpack/assets/javascripts/bundle.js [hunk 1]
```diff
@@ -4,6 +4,13 @@
 /* eslint-disable import/no-unresolved */
 /* eslint-disable import/extensions */
 /* eslint-disable import/first */
+
+// Set the public path for dynamic imports
+if (process.env.NODE_ENV !== 'production') {
```
Comment (ares) [inline webpack/assets/javascripts/bundle.js:9]: with this, we can no longer test production like compilation through dev server, but I guess it's not a big deal

**Review by ares — COMMENTED** [review]
I was able to make it work when I disabled production NODE_ENV and allowed configuring host and port for manifest download. I opened https://github.com/theforeman/foreman/pull/6378 which should cause no problems to existing setups if I set the right defaults.

**Pull request merged**